openrct2-ports/icu/portfile.cmake

```diff
@@ -0,0 +1,311 @@
+        message(STATUS "Configuring ${RELEASE_TRIPLET} done")
+    endif()
+
+    if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
+        # Configure debug
+        message(STATUS "Configuring ${DEBUG_TRIPLET}")
+        file(REMOVE_RECURSE ${CURRENT_BUILDTREES_DIR}/${DEBUG_TRIPLET})
+        file(MAKE_DIRECTORY ${CURRENT_BUILDTREES_DIR}/${DEBUG_TRIPLET})
+        set(ENV{CFLAGS} "${EXTRA_DEBUG_FLAGS} ${VCPKG_C_FLAGS} ${VCPKG_C_FLAGS_DEBUG}")
+        set(ENV{CXXFLAGS} "${EXTRA_DEBUG_FLAGS} ${VCPKG_CXX_FLAGS} ${VCPKG_CXX_FLAGS_DEBUG}")
+        set(ENV{LDFLAGS} "${DEBUG_LDFLAGS}")
+        vcpkg_execute_required_process(
+            COMMAND ${BASH} --noprofile --norc -c
+                "${SOURCE_PATH}/source/runConfigureICU ${PLATFORM} ${CONFIGURE_OPTIONS} ${CONFIGURE_OPTIONS_DEBUG}"
+            WORKING_DIRECTORY "${CURRENT_BUILDTREES_DIR}/${DEBUG_TRIPLET}"
+            LOGNAME "configure-${DEBUG_TRIPLET}")
+        message(STATUS "Configuring ${DEBUG_TRIPLET} done")
+    endif()
+endif()
+
+unset(ENV{CFLAGS})
+unset(ENV{CXXFLAGS})
+unset(ENV{LDFLAGS})
+
+if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
+    # Build release
+    message(STATUS "Package ${RELEASE_TRIPLET}")
+    vcpkg_execute_build_process(
+        COMMAND ${BASH} --noprofile --norc -c "make -j ${VCPKG_CONCURRENCY}"
+        NO_PARALLEL_COMMAND ${BASH} --noprofile --norc -c "make"
+        WORKING_DIRECTORY "${CURRENT_BUILDTREES_DIR}/${RELEASE_TRIPLET}"
+        LOGNAME "make-build-${RELEASE_TRIPLET}")
+
+    if(VCPKG_TARGET_IS_OSX AND VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
+        if(DEFINED CMAKE_INSTALL_NAME_DIR)
+            set(ID_PREFIX "${CMAKE_INSTALL_NAME_DIR}")
+        else()
+            set(ID_PREFIX "@rpath")
+        endif()
+        
+        vcpkg_execute_build_process(
+            COMMAND /usr/bin/install_name_tool -id "${ID_PREFIX}/libicuuc.${ICU_VERSION_MAJOR}.dylib"
+            "libicuuc.${VERSION}.dylib"
+            WORKING_DIRECTORY "${CURRENT_BUILDTREES_DIR}/${RELEASE_TRIPLET}/lib"
+            LOGNAME "make-build-fix-rpath-${RELEASE_TRIPLET}"
+        )
+
+        vcpkg_execute_build_process(
+            COMMAND /usr/bin/install_name_tool -id "${ID_PREFIX}/libicudata.${ICU_VERSION_MAJOR}.dylib"
+            "libicudata.${VERSION}.dylib"
+            WORKING_DIRECTORY "${CURRENT_BUILDTREES_DIR}/${RELEASE_TRIPLET}/lib"
+            LOGNAME "make-build-fix-rpath-${RELEASE_TRIPLET}"
+        )
+
+        vcpkg_execute_build_process(
+            COMMAND /usr/bin/install_name_tool -change "libicudata.${ICU_VERSION_MAJOR}.dylib"
+                                                   "${ID_PREFIX}/libicudata.${ICU_VERSION_MAJOR}.dylib"
+                                                   "libicuuc.${VERSION}.dylib"
+            WORKING_DIRECTORY "${CURRENT_BUILDTREES_DIR}/${RELEASE_TRIPLET}/lib"
+            LOGNAME "make-build-fix-rpath-${RELEASE_TRIPLET}"
+        )
+    endif()
```

Comment:
This is the block of changes from the main icu port. We could also do this as a patch file, similar to https://github.com/OpenRCT2/Dependencies/blob/master/0001-Patch-duktape.patch, and add the patch application to the build. Let me know if that would be preferred and I'll switch this over to go that route instead of overlay-ports.

Reply:
Is this a change that can be submitted to the upstream vcpkg project. I'd be hesitant to use a patch or custom port for this sort of thing as vcpkg changes often, and other users could benefit from our fixes. I usually find that PRs can be merged within about a week if you keep up with requests they make.

Reply:
I only added the fixes for the two icu libraries we use - the upstream port would need to patch all of them. I wasn't entirely sure if this was even the best way to patch them, or if the icu configure scripts for --enable-rpath should be enhanced. I'm still learning the ins and outs of rpath, so I figured it'd be better to patch it ourselves, figure out what to do upstream, and if upstream merges a change, we can remove our patch.

As you can probably tell, I modeled this after your openssl port changes, so perhaps changing this after the build like this is ok? I'll just need to add a loop and cover all of the icu modules.

Reply:
This looks to be the issue upstream, btw:
https://github.com/unicode-org/icu/blob/master/icu4c/source/config/mh-darwin#L32-L37
With --enable-rpath on, the install dir is used instead of @rpath. I tried passing `-libdir=@rpath` to configure, but it expects an absolute path for libdir if it is passed.

Reply:
If this is an upstream issue, then a) Making a patch to the icu project might be a good idea (probably). b) This seems like a good opportunity for a port patch. In [the icu port directory](https://github.com/microsoft/vcpkg/tree/master/ports/icu) there are a bunch of .patch files that get applied to the base project in `portfile.cmake`. If it's a matter of just changing `-install_name $(libdir)/$(notdir $(MIDDLE_SO_TARGET)` to `-install_name @rpath/$(notdir $(MIDDLE_SO_TARGET)`, then that's something to consider instead of changing the install-name post-build (which isn't an optimal solution).

Reply:
Wouldn't `VCPKG_INSTALL_NAME_DIR` be passed to icu as the `$(libdir)` already? I believe that behavior is correct already - when you pass `--enable-rpath`, the library is built with the absolute path to the library directory as its path. What we need is a new option for using the rpath prefix. I'll have to add it to the icu configure script to make it accessible at the point where we'd want to make that change. Good idea on a patch though. Let me give it a try.

The ICU project seems to have some thoughts on rpath ;) https://github.com/unicode-org/icu/blob/master/icu4c/source/configure#L6065

Reply:
`VCPKG_INSTALL_NAME_DIR` is distinct from what is probably `$(libdir)`. Setting `VCPKG_INSTALL_NAME_DIR` in a triplet changes the path in the install-name to a different specified path from where it's installed to. So if it is installing `/some/install/path/libx.dylib`, but the install-name should be `/some/other/path/because/reasons/libx.dylib`, then you can add to the triplet file `set(VCPKG_INSTALL_NAME_DIR "/some/other/path/because/reasons")`. This variable basically just sets [`CMAKE_INSTALL_NAME_DIR`](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_NAME_DIR.html). The default is to use `@rpath` thanks to another CMAKE variable.